packages/crypto/crypto-rs/README.md

```diff
@@ -1,94 +1,61 @@
-# iroha-crypto Rust sources
+# Iroha Crypto WASM (Rust sources)
 
-## Interfaces rules
+This Cargo project is a port of [`iroha_crypto` crate](https://github.com/hyperledger/iroha/tree/iroha2-lts/crypto) with `wasm_bindgen`s.
 
-**Don't expose ownership model implicitly**
+## Rebuild WASMs
 
-It is not visible from JS side. Prefer always to borrow structs.
+To re-create `./wasm-pkg-*`, run in the monorepo root:
 
-```rust
-#[wasm_bindgen]
-pub struct User {
-    name: String
-}
+```bash
+pnpm jake crypto-wasm:rebuild
+```
 
-// WRONG
-// `user` will be freed after invocation of this function so
-// any usage of related object on JS side after this will be
-// incorrect
-#[wasm_bindgen]
-pub fn get_user_name(user: User) -> String {
-    user.name
-}
+## Be aware of moves
 
-// GOOD
-#[wasm_bindgen]
-pub fn get_user_name(user: &User) -> String {
-    user.name.clone()
-}
-```
+Take a look at this Rust code:
 
-**Avoid "static" class methods**
+```rs
+#[wasm_bindgen]
+struct Foo(u32);
 
-Classes are the thing that is hard to type in TypeScript. When you write code like this:
+#[wasm_bindgen]
+struct Bar(u32);
 
-```rust
 #[wasm_bindgen]
-pub struct User {
-    name: String
+fn create_bar() -> Bar {
+    Bar(42)
 }
 
 #[wasm_bindgen]
-impl User {
-    #[wasm_bindgen(constructor)]
-    pub fn new(name: String) -> User {
-        User { name }
-    }
-
-    pub fn with_empty_name() -> User {
-        User::new("".to_owned())
-    }
-
-    pub fn get_name(&self) -> String {
-        self.name.clone()
-    }
+fn foo_from_bar(bar: Bar) -> Foo {
+    Foo(bar.0 - 16)
 }
 ```
 
-It transforms into such class:
+The problem appears when you use this code (compiled to WASM) in JavaScript:
 
-```ts
-declare class User {
-    constructor(name: string);
+```js
+const bar = create_bar()
 
-    static with_empty_name(): User;
+const foo = foo_from_bar(bar)
+//                       ^^^ `bar` is moved here
 
-    get_name(): string;
-}
+// panic! null ptr is passed to Rust
+const foo2 = foo_from_bar(bar)
+//                        ^^^ `bar` cannot be used second time
 ```
 
-The problem is that library have to expose universal API interface for each target and it is decoupled from implementation, but classes are always a mix of runtime and types sides. Also it has 2 type sides: static type and instance type. It is hard to pack it all in `IrohaCryptoInterface`, so I decided to move all static methods away from classes and leave there only instance methods. So, do like this:
+Code panics, because `bar` is used **after it was moved to `foo_from_bar`**. It is completely valid from Rust side and such a usage of `bar` after its movement will cause a compilation error. However, JavaScript can't apply borrowing rules on its side.
```

Comment:
Hm, now I think that saying "it is completely valid" is weird, because it is not valid - it is a compilation error. Maybe, this is better?

> From Rust side, using `bar` after moving it to `foo_from_bar` causes compilation error. However, from JavaScript side there is no way to prevent violation of borrowing rules.

Reply:
slight modifications:
> In Rust, using `bar` after moving it to `foo_from_bar` causes compilation error. However, in JavaScript there is no way to prevent violation of borrowing rules.